test/test_selectors.py

```diff
@@ -0,0 +1,622 @@
+        s.register(rd, selectors.EVENT_READ)
+        key = s.get_key(rd)
+
+        def second_alarm(*args):
+            wr.send(b'x')
+
+        def first_alarm(*args):
+            signal.alarm(0)
+            signal.alarm(1)
+            signal.signal(signal.SIGALRM, second_alarm)
+
+        sigalrm_handler = signal.signal(signal.SIGALRM, first_alarm)
+        self.addCleanup(signal.signal, signal.SIGALRM, sigalrm_handler)
+        self.addCleanup(signal.alarm, 0)
+
+        # Start the timer for the interrupt.
+        signal.alarm(1)
+
+        t = monotonic()
+        self.assertEqual([(key, selectors.EVENT_READ)], s.select(3))
+        self.assertLess(monotonic() - t, 3.2)
+        self.assertEqual(rd.recv(1), b'x')
+
+    # This is to ensure that the _syscall_wrapper still follows
+    # specification from PEP 475 which is that any system call
+    # that is interrupted and the interrupt handler raises an
+    # exception should not retry the system call and instead
+    # raise the exception.
+    @skipUnless(hasattr(signal, "alarm"), "Platform doesn't have signal.alarm()")
+    def test_select_interrupt_exception(self):
+        s = self.make_selector()
+        rd, wr = self.make_socketpair()
+        s.register(rd, selectors.EVENT_READ)
+
+        class AlarmInterrupt(Exception):
+            pass
+
+        def alarm_exception(*args):
+            raise AlarmInterrupt()
+
+        sigalrm_handler = signal.signal(signal.SIGALRM, alarm_exception)
+        self.addCleanup(signal.signal, signal.SIGALRM, sigalrm_handler)
+        self.addCleanup(signal.alarm, 0)
+
+        # Start the timer for the interrupt.
+        signal.alarm(1)
+
+        t = monotonic()
+        self.assertRaises(AlarmInterrupt, s.select, 2.0)
+        self.assertLess(monotonic() - t, 1.2)
+
+    def test_fileno(self):
+        s = self.make_selector()
+        if hasattr(s, "fileno"):
+            fd = s.fileno()
+            self.assertTrue(isinstance(fd, int))
+            self.assertGreaterEqual(fd, 0)
+
+    # According to the psutil docs, open_files() has strange behavior
+    # on Windows including giving back incorrect results so to
+    # stop random failures from occurring we're skipping on Windows.
+    @skipIf(sys.platform == "win32", "psutil.Process.open_files() is unstable on Windows.")
+    def test_leaking_fds(self):
+        proc = psutil.Process()
+        before_fds = len(proc.open_files())
+        s = self.make_selector()
+        s.close()
+        after_fds = len(proc.open_files())
+        self.assertEqual(before_fds, after_fds)
+
+    @skipUnless(sys.platform == "win32", "This is the same as test_leaking_fds but for Windows")
+    def test_leaking_fds_windows(self):
+        proc = psutil.Process()
+        before_fds = len(proc.open_files())
+        s = self.make_selector()
+        s.close()
+        after_fds = len(proc.open_files())
+        if before_fds != after_fds:
+            self.skipTest("Leaked fd but test was on Windows.")
```

Comment:
I don't think that we can rely on that. Developers won't spot that "skip means fail".